### PR TITLE
Remove the CST Visitor API

### PR DESCRIPTION
**Pull request by Xanewok**
We discussed internally that the Visitor API for (dynamically typed) CST is a lot less useful now that we have the `Iterator` interface implemented in Rust (and hopefully [`Iterator`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Iterators_and_Generators#iterators) protocol in TS in the future), so let's remove the CST Visitor API altogether.

The Visitor [pattern](https://rust-unofficial.github.io/patterns/patterns/behavioural/visitor.html) will be a lot useful once we have typed, heterogenous data such as the new AST (#634), so let's wait until then with designing and exposing such interface.

### CODE REVIEW

File: crates/solidity/testing/snapshots/cst_output/SourceUnit/everything/generated/0.4.11-failure.yml [hunk 1]
```diff
@@ -60,5 +60,5 @@ Tree:
               - PathImport (Rule): # 30..40 ' "foo.sol"'
                   - AsciiStringLiteral (Token): '"foo.sol"' # 31..40
               - Semicolon (Token): ";" # 40..41
-      - EndOfFileTrivia (Rule): "\n" # 42..43
+      - EndOfFileTrivia (Rule): # 42..43 "\n"
```
Comment (Xanewok) [inline crates/solidity/testing/snapshots/cst_output/SourceUnit/everything/generated/0.4.11-failure.yml:63]: Previously, the `TestNodeBuilder` would not construct the whitespace children tokens and the printer logic would recognize this as an childless node, whereas the new printer does not, as we only skipped the leading/trailing trivia nodes before.

This should probably be fixed to also not include the `EndOfFileTrivia` much like the Leading/Trailing variants, however this is not as important as we will change the trivia model soon (DSL v2).

Comment (Xanewok) [reply crates/solidity/testing/snapshots/cst_output/SourceUnit/everything/generated/0.4.11-failure.yml]: See 47663c9313726734550b4c87bbef2208f1d5bf0b for the fix

**Review by OmarTawfik — COMMENTED** [review]
left a few suggestions.

File: crates/codegen/parser/runtime/src/visitor.rs [hunk 2]
```diff
@@ -1,76 +1 @@
-use std::ops::ControlFlow;
-use std::rc::Rc;
```
[the remark this answers is not in the document]
Comment (Xanewok) [reply crates/codegen/parser/runtime/src/visitor.rs:2]: Ah, good catch! Let me fix that.

File: crates/codegen/parser/runtime/src/cst.rs [hunk 3]
```diff
@@ -46,6 +46,13 @@ impl Node {
         }
     }
```
[the remark this answers is not in the document]
Comment (Xanewok) [reply crates/codegen/parser/runtime/src/cst.rs:48]: None of these functions have a doc comment, so I wanted to stay consistent and this function is trivial but I will add a doc comment if you'd like

File: crates/codegen/parser/runtime/src/cst.rs [hunk 4]
```diff
@@ -46,6 +46,13 @@ impl Node {
...
 
+    pub fn children(&self) -> &[Node] {
```
[the remark this answers is not in the document]
Comment (Xanewok) [reply crates/codegen/parser/runtime/src/cst.rs:50]: As always, it depends. If we don't want to commit to internals in the public API and don't want to expose them (e.g. specific key-value collection) or we want to use some generic combinators (e.g. filters, maps) and don't want to create a dedicated public named type for that, it's more appropriate to use an `impl Iterator`.

However, in many cases it's fine to return a slice - there's not much other ways we can implement this in this case, the `Node` type is relatively stable and simple enough and I wanted the `is_empty()` method (not really a concern and it's only called in the CST/test code but still).

File: crates/codegen/parser/runtime/src/cursor.rs [hunk 5]
```diff
@@ -144,11 +144,13 @@ impl Cursor {
         self.current.text_range()
     }
 
-    pub fn path_rule_nodes(&self) -> Vec<Rc<RuleNode>> {
-        self.path
-            .iter()
-            .map(|path_element| path_element.rule_node.clone())
-            .collect()
+    pub fn depth(&self) -> usize {
```
[the remark this answers is not in the document]
Comment (Xanewok) [reply crates/codegen/parser/runtime/src/cursor.rs:148]: done

**Review by OmarTawfik — APPROVED** [review]

**Pull request merged**